README.md
# ttwidgets

Package **TTWidgets** improves the `Button`, `Label`, and `Listbox` widgets of
the Tkinter library and provides a new `ToolTip` widget, all with support for
multiple fonts and visual schemes.

Although the standard Tkinter `Button` and `Label` widgets are limited to a
single set of widget options (e.g. background/bg, foreground/fg, bitmap,
cursor, relief) and a **single font**, a **TTWidgets** enhanced widget can be
passed *tagged text* instead of plain text to create a visually compound widget 
(with multiple fonts and visual option sets) that behaves like a simple widget.

- The `Button` and `Label` implementations are complete.
- The `Listbox` implementation is partial: there is no multi-font support, but 
    the user can use the new 'text' option to pass in *tagged text* to define all
    the elements with visual schemes.  The user can also pass *tagged text* to the
    insert() method, and thereby skip the secondary call to itemconfig() for the
    inserted element(s).  A full multi-font implementation of a *tagged text*
    `Listbox` may come later.
- As a bonus, a `ToolTip` widget is included, which also accepts *tagged text*
    as input, allowing the creation of colorful and multi-font ToolTips.

All **TTWidgets** support the 'text' keyword to accept and process *tagged text*
as input on creation. 
All **TTWidgets** support modification of 'text' via the `config()` method and
retrieval via `cget()`.

As a **ttwidgets** Button example:

    button = ttwidgets.TTButton(text="Isn't a <t case=title relief=raised bd=1 bg=white padx=2>button</t> with <tag bold fg=red bg=yellow>bold red on yellow text</tag> and \na <tag fg=blue funderline>hyperlink</tag> <tag relief=groove bd=2>groovy</tag> <tag bitmap=warning/>", command=lambda e=None: print("Released!"), bg='lightgray')

gives a Button with text "Isn't a Button with bold red on yellow text and
        a hyperlink groovy!", where:
- "Button" appears inside a raised button (with "title" case),
- "bold red on yellow text" is bold red text on a yellow background,
- "hyperlink" is blue and underlined,
- "groovy" is in a box with 'groove' relief, and
- "!" is a warning bitmap.

Here are two versions of the same button.  Which one would YOU rather use? :)

![exampleButton](https://user-images.githubusercontent.com/19311746/75097457-fa2e5b80-5578-11ea-9be4-016763568e50.jpg "Example TTButton")

As a **ttwidgets** ToolTip example:

    tooltip = ttwidgets.TTToolTip(reg_but, text='My <t bg=cyan px=2 py=2 bd=2 relief=groove size=6 wraplength=70>Regular Button Cycle All States</t><t b> ToolTip</t>... See how it matches? :)', bg='white', relief=tk.GROOVE, borderwidth=5)
    
gives a ToolTip for button "reg_but" which includes a smaller version of the
button it is tipping, to make the ToolTip more visually appealing. The ToolTip
also shows bold text and a white background.

![exampleToolTip](https://user-images.githubusercontent.com/19311746/75097456-fa2e5b80-5578-11ea-8768-bce0c4306874.jpg "Example TTToolTip")

In general, all the visual options and font attributes are supported in the
*tagged text*.  For a *tagged text* overview, please see **ttwidgets** class `TTWidget` help.


CHANGES.txt
v1.0.0, Mar 01, 2020 -- Initial release.
v1.0.1, Mar 02, 2020 -- Updated with code on GitHub
v1.0.2, Mar 02, 2020 -- Updated Package URL and supported Python versions
v1.0.3, Mar 04, 2020 -- Fixes for Linux Test
v1.0.4, Mar 09, 2020 -- More Fixes for Linux and Darwin. Dropped Py3.4 support.
v1.0.5, Mar 14, 2020 -- Reworked ToolTip.

ttwidgets/test/test_ttwidgets.py
"""
test_ttwidgets
==============

This runs a quick test of the expected output formatting from the module
to make sure it is installed and running.
"""

"""
Copyright 2020 Gary Michael Bloom

               mailto:bloominator@hotmail.com
               mailto:GaryBloomLaw@gmail.com

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
"""

import tkinter as tk
import unittest
import ttwidgets


class Test_PyVers(unittest.TestCase):

    def test_pyvers(self):
        self.assertTrue(ttwidgets.PyVers_f >= 3.1,
                        "Unsupported Python Version %s" % ttwidgets.PyVers_s)


class Test_fromDicts(unittest.TestCase):
    """Test from dicts."""

    def test_get_tag_attrs_fromDicts1(self):
        o_d = dict(family="Courier New", background='white', fg='blue', bd='2',
                   relief=tk.SOLID, funderline=1, foverstrike=0, underline=10)
        gen_text = ttwidgets.gen_tag_attrs(None, o_d, kmode='')
        # abbrev=False) # , debug=True)
        expected_text = 'family="Courier New" background=white fg=blue bd=2' \
                        ' relief=solid funderline=1 foverstrike=0 underline=10'
        if ttwidgets.PyVers_f < 3.6:  # False: # True: #
            gen_text = sorted(ttwidgets.split_attrs(gen_text))
            expected_text = sorted(ttwidgets.split_attrs(expected_text))
        self.assertEqual(gen_text, expected_text)

    def test_get_tag_attrs_fromDicts2(self):
        o_d = dict(family='Courier New', background='white', fg='blue', bd='2',
                   relief=tk.SOLID, funderline=1, foverstrike=0, underline=10)
        gen_text = ttwidgets.gen_tag_attrs(options=o_d, kmode='alias')
        # abbrev=True) # , debug=True)
        expected_text = 'fam="Courier New" bg=white fg=blue bd=2 rel=solid ' \
                        'u=1 o=0 ul=10'
        if ttwidgets.PyVers_f < 3.6:  # False: # True: #
            gen_text = sorted(ttwidgets.split_attrs(gen_text))
            expected_text = sorted(ttwidgets.split_attrs(expected_text))
        self.assertEqual(gen_text, expected_text)

    def test_get_tag_attrs_fromDicts3(self):
        o_d = dict(family='Courier New', background='white', fg='blue', bd='2',
                   relief=tk.SOLID, funderline=1, foverstrike=0, underline=10)
        gen_text = ttwidgets.gen_tag_attrs(None, o_d, kmode='option')
        # abbrev=True) # , debug=True)
        expected_text = 'family="Courier New" background=white foreground=' \
                        'blue borderwidth=2 relief=solid funderline=1 ' \
                        'foverstrike=0 underline=10'
        if ttwidgets.PyVers_f < 3.6:  # False: # True: #
            gen_text = sorted(ttwidgets.split_attrs(gen_text))
            expected_text = sorted(ttwidgets.split_attrs(expected_text))
        self.assertEqual(gen_text, expected_text)

    def test_get_tag_attrs_fromDicts4(self):
        o_d = dict(background='white', fg='blue', bd='2', relief=tk.SOLID,
                   underline=10)
        f_d = dict(family='Courier New', underline=1, overstrike=0, )
        gen_text = ttwidgets.gen_tag_attrs(None, o_d, f_d,
                                           kmode='option')
        # abbrev=True) # , debug=True)
        expected_text = 'background=white foreground=blue borderwidth=2 ' \
                        'relief=solid family="Courier New" funderline=1 ' \
                        'foverstrike=0 underline=10'
        if True:  # ttwidgets.PyVers_f < 3.6: # False: #
            gen_text = sorted(ttwidgets.split_attrs(gen_text))
            expected_text = sorted(ttwidgets.split_attrs(expected_text))
        self.assertEqual(gen_text, expected_text)

    def test_get_tag_attrs_fromDicts5(self):
        o_d = dict(background='white', foreground='blue', borderwidth='2',
                   relief=tk.SOLID, underline=10)
        f_d = dict(family='Courier New', underline=1, overstrike=0, )
        case = 'title'
        gen_text = ttwidgets.gen_tag_attrs(None, o_d, f_d, case,
                                           kmode='option')
        #                                  abbrev=True) # , debug=True)
        expected_text = 'background=white foreground=blue borderwidth=2 ' \
                        'relief=solid family="Courier New" funderline=1 ' \
                        'foverstrike=0 ' \
                        'underline=10 case=title'
        o_exp_d, f_exp_d, case_exp = ttwidgets.parse_tag_attrs(expected_text)
        if True:  # ttwidgets.PyVers_f < 3.6: # False: #
            gen_text = sorted(ttwidgets.split_attrs(gen_text))
            expected_text = sorted(ttwidgets.split_attrs(expected_text))
        self.assertEqual(gen_text, expected_text)
        self.assertEqual(o_d, o_exp_d)
        self.assertEqual(f_d, f_exp_d)
        self.assertEqual(case, case_exp)


class Test_fromWidget(unittest.TestCase):
    """Test from widget."""

    def test_get_tag_attrs_fromWidget1(self):
        but = ttwidgets.Button(bg='wheat',
                               text='<tag fg=red>red</tag> and <tag'
                               ' background=blue>blue</tag>')
        gen_text = but.gen_tag_attrs(kmode='', extend=True)
        # abbrev=True) # , debug=True)
        expected_text = ['background=wheat text="<tag fg=red>red</tag> and '
                         '<tag background=blue>blue</tag>"',
                         'foreground=red text=red',
                         'text=" and "',
                         'background=blue text=blue']
        if ttwidgets.PyVers_f < 3.6:  # False: # True: # False: #
            gen_text = \
                [sorted(ttwidgets.split_attrs(elem)) for elem in gen_text]
            expected_text = \
                [sorted(ttwidgets.split_attrs(elem)) for elem in expected_text]
        self.assertEqual(gen_text, expected_text)

    def test_get_tag_attrs_fromWidget2(self):
        # toplevel = tk.Tk()
        but = ttwidgets.Button(bg='wheat',
                               text='<tag fg=red>red</tag> and '
                               '<tag background=blue>blue</tag>')
        gen_text = but.gen_tag_attrs(kmode='alias', extend=True)
        # abbrev=True) # , debug=True)
        expected_text = ['bg=wheat txt="<tag fg=red>red</tag> and <tag'
                         ' background=blue>blue</tag>"',
                         'fg=red txt=red',
                         'txt=" and "',
                         'bg=blue txt=blue']
        if ttwidgets.PyVers_f < 3.6:  # False: # True: # False: #
            gen_text = \
                [sorted(ttwidgets.split_attrs(elem)) for elem in gen_text]
            expected_text = \
                [sorted(ttwidgets.split_attrs(elem)) for elem in expected_text]
        self.assertEqual(gen_text, expected_text)

    def test_get_tag_attrs_fromWidget3(self):
        # toplevel = tk.Tk()
        but = ttwidgets.Button(bg='wheat',
                               text='<tag fg=red>red</tag> and <tag '
                               'background=blue>blue</tag>')
        gen_text = but.gen_tag_attrs(kmode='option', extend=True)
        # abbrev=True) # , debug=True)
        expected_text = ['background=wheat text="<tag fg=red>red</tag> and '
                         '<tag background=blue>blue</tag>"',
                         'foreground=red text=red',
                         'text=" and "',
                         'background=blue text=blue']
        if ttwidgets.PyVers_f < 3.6:  # False: # True: # False: #
            gen_text = \
                [sorted(ttwidgets.split_attrs(elem)) for elem in gen_text]
            expected_text = \
                [sorted(ttwidgets.split_attrs(elem)) for elem in expected_text]
        self.assertEqual(gen_text, expected_text)

    def test_get_tag_attrs_fromWidget4(self):
        # toplevel = tk.Tk()
        but = ttwidgets.Button(bg='wheat',
                               text='<tag fg=red>red</tag> and <tag'
                               ' background=blue>blue</tag>')
        gen_text = but.gen_tag_attrs(kmode='option', extend=False)
        # abbrev=True) # , debug=True)
        expected_text = ['background=wheat text="<tag fg=red>red</tag> '
                         'and <tag background=blue>blue</tag>"',
                         'foreground=red', '', 'background=blue']
        # print('\n\n', pprint.pformat(gen_text))
        # print('\n\n', pprint.pformat(expected_text))
        if ttwidgets.PyVers_f < 3.6:  # False: # True: # False: #
            gen_text = \
                [sorted(ttwidgets.split_attrs(elem)) for elem in gen_text]
            expected_text = \
                [sorted(ttwidgets.split_attrs(elem)) for elem in expected_text]
        self.assertEqual(gen_text, expected_text)


if __name__ == '__main__':
    unittest.main()


ttwidgets/__init__.py
"""
   Copyright 2020 Gary Michael Bloom
                  mailto:bloominator@hotmail.com
                  mailto:GaryBloomLaw@gmail.com

TTWidgets - Tagged-Text Multi-Font Enhanced Tkinter Widgets
=========

Package **TTWidgets** improves the Button, Label, and Listbox widgets of the
Tkinter library, and provides a bonus ToolTip widget, all with support for 
multiple fonts and visual schemes via *tagged-text* input.

The standard Tkinter Button and Label widgets are limited to a single set
of widget options (e.g. background/bg, foreground/fg, bitmap, cursor,
relief, etc), including a SINGLE FONT. With the **TTWidgets** enhancements,
*tagged text* can be passed in as text to create a compound widget, with
multiple fonts and visual option sets, that behaves like a simple widget.

The Button and Label implementations are complete.
The Listbox implementation is partial: there is no multi-font support, but 
the user can use the new 'text' option to pass in *tagged text* to define all
the elements with visual schemes.  The user can also pass *tagged text* to the
insert() method, and thereby skip the secondary call to itemconfig() for the
inserted element(s).
A full multi-font implementation of a Listbox may come later.

As a bonus, a ToolTip widget is included, which also accepts *tagged text*
as an input option, allowing the creation of colorful and multi-font ToolTips.

As a **TTWidgets** Button example:

    button = ttwidgets.TTButton(text="Isn't a <t case=title relief=raised bd=1 bg=white padx=2>button</t> with <tag bold fg=red bg=yellow>bold red on yellow text</tag> and \na <tag fg=blue funderline>hyperlink</tag> <tag relief=groove bd=2>groovy</tag> <tag bitmap=warning/>", command=lambda e=None: print("Released!"), bg='lightgray')

gives a Button with text "Isn't a Button with bold red on yellow text and
        a hyperlink groovy!", where:
- "Button" appears inside a raised button (with "title" case),
- "bold red on yellow text" is bold red text on a yellow background,
- "hyperlink" is blue and underlined,
- "groovy" is in a box with 'groove' relief, and
- "!" is a warning bitmap.

Here are two versions of the same button.  Which one would YOU rather use? :)

![exampleButton](https://user-images.githubusercontent.com/19311746/75097457-fa2e5b80-5578-11ea-9be4-016763568e50.jpg "Example TTButton")

For an overview of *tagged text*, please see the ttwidgets.TTWidget help.

This package includes the following:
    CLASSES:
        TTWidget    (base class for TTButton and TTLabel)
        TTButton    (inherits from TTWidget)
        TTLabel     (inherits from TTWidget)
        TTListbox   (inherits from Tkinter.Listbox)
        TTToolTip   (does not inherit, but uses a TTLabel)
    METHODS:    
        alias
        convert_font_dict_to_ttoptions_dict
        convert_ttoptions_dict_to_font_dict
        dump
        gen_tag_attrs
        get_font_dict
        get_named_font
        is_tagged_text
        pare_dict
        parse_tag_attrs
        quote
        split_attrs
        split_chunk
        split_dict_into_options_fontattrs_and_case
        split_tagged_text_into_chunks
        strip_tags
        unalias
        unmap
        update_named_font
        wrap_tagged_text


    Compatability:
        Py 3.5 - 3.9 is supported.
        Py 3.4 and earlier are unsupported

Copyright 2020 Gary Michael Bloom

               mailto:bloominator@hotmail.com
               mailto:GaryBloomLaw@gmail.com

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
"""

from .ttwidgets import *

__version__ = "1.0.5"
